<launch>
<arg name="debug" default="false"/>
<arg name="rviz" default="false"/>
<arg name="gazebo_gui" default="false"/>
<arg name="TAO" default="false"/>
<arg name="sim" default="false"/>
<arg name="start_paused" default="false"/>
<arg name="use_sim_time" default="true"/>
<arg name="root_link" default="link_0"/>
<arg name="tip_link" default="link_7"/>
<!-- Tools -->
<arg name="load_base" default="false" />
<arg name="load_ati_sensor" default="false" />
<arg name="load_calib_tool" default="false" />
<arg name="load_table" default="false" />
<arg name="load_handle" default="false" />
<!-- Start Gazebo here or start it elsewhere and leave it open -->
<arg if="$(arg sim)"     name="launch_gazebo" default="true" />
<arg unless="$(arg sim)" name="launch_gazebo" default="false" />

<arg if="$(arg sim)"     name="robot_name" default="lwr_sim"/>
<arg unless="$(arg sim)" name="robot_name" default="lwr"/>

<arg if="$(arg sim)"     name="robot_ns" default="/" />
<arg unless="$(arg sim)" name="robot_ns" default="/" />

<arg name="tf_prefix" default="$(arg robot_ns)"/>

<arg if="$(arg sim)"     name="rviz_config" default="$(find lwr_utils)/launch/lwr_sim.rviz"/>
<arg unless="$(arg sim)" name="rviz_config" default="$(find lwr_utils)/launch/lwr_hw.rviz"/>

<node if="$(arg rviz)" name="rviz" pkg="rviz" type="rviz" args="-d $(arg rviz_config)"/>

<arg name="initial_joint_positions" default="[0, 0, 0, 0, 0, 0, 0]"/>

<group if="$(arg use_sim_time)">
  <param name="/use_sim_time" value="true" />
</group>

<!-- Description and/or gazebo -->
<group ns="$(arg robot_ns)">
    <!--group ns="$(arg robot_name)"-->
    <param name="tf_prefix" type="str" value="$(arg tf_prefix)"/>
    <param name="robot_name" type="str" value="$(arg robot_name)"/>
    <param name="robot_ns" type="str" value="$(arg robot_ns)"/>
    <param name="root_link" type="str" value="$(arg root_link)"/>
    <param name="tip_link" type="str" value="$(arg tip_link)"/>
    <param name="is_sim" type="bool" value="$(arg sim)"/>

    <!-- Gazebo launches it if sim, otherwise here so that lwr_gazebo is standalone (hacky) -->
    <include file="$(find lwr_description)/launch/lwr_upload.launch">
        <arg name="load_base" value="$(arg load_base)" />
        <arg name="load_ati_sensor" value="$(arg load_ati_sensor)" />
        <arg name="load_calib_tool" value="$(arg load_calib_tool)" />
        <arg name="load_table" value="$(arg load_table)" />
        <arg name="load_handle" value="$(arg load_handle)" />
        <arg name="robot_name" value="$(arg robot_name)" />
    </include>

    <!-- world is in the urdf, but sometime we want everything in a different namespace, so we'll have /sub/world. This connects /world to /sub/world-->
    <node pkg="tf" type="static_transform_publisher" name="world_to_$(arg robot_name)_world" args="0 0 0 0 0 0 /world $(arg tf_prefix)/world 100"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
        <param name="publish_frequency" value="1000"/>
    </node>
    <!--/group-->
</group>

<node if="$(arg gazebo_gui)" name="gazebo_gui" pkg="gazebo_ros" type="gzclient" respawn="false" output="screen"/>
<!--include if="$(arg launch_gazebo)" file="$(find lwr_utils)/launch/lwr_gazebo.launch">
    <arg name="robot_ns" value="$(arg robot_ns)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="robot_name" value="lwr_sim"/>
    <arg name="root_link" value="$(arg root_link)"/>
    <arg name="tip_link" value="$(arg tip_link)"/>
    <arg name="gui" value="$(arg gazebo_gui)"/>
    <arg name="load_robot_description" value="false"/>
    <arg name="load_base" value="$(arg load_base)" />
    <arg name="load_ati_sensor" value="$(arg load_ati_sensor)" />
    <arg name="load_calib_tool" value="$(arg load_calib_tool)" />
    <arg name="load_table" value="$(arg load_table)" />
    <arg name="load_handle" value="$(arg load_handle)" />
    <arg name="paused" value="$(arg start_paused)"/>
    <arg name="initial_joint_positions" value="$(arg initial_joint_positions)"/>
    <arg name="use_sim_time" value="$(arg use_sim_time)"/>
</include-->
<include if="$(arg launch_gazebo)" file="$(find lwr_utils)/launch/spawn_robot.launch">
    <arg name="robot_ns" value="$(arg robot_ns)"/>
    <arg name="robot_name" value="$(arg robot_name)"/>
    <arg name="x" value="0.0"/>
    <arg name="y" value="0.0"/>
    <arg name="z" value="0.0"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="root_link" value="$(arg root_link)"/>
    <arg name="tip_link" value="$(arg tip_link)"/>
    <arg name="load_base" default="$(arg load_base)" />
    <arg name="load_ati_sensor" default="$(arg load_ati_sensor)" />
    <arg name="load_calib_tool" default="$(arg load_calib_tool)" />
    <arg name="load_table" default="$(arg load_table)" />
    <arg name="load_handle" default="$(arg load_handle)" />
    <arg name="initial_joint_positions" default="$(arg initial_joint_positions)" />
</include>
</launch>
